pkg/proxy/ipvs/meta-proxier.go

```diff
@@ -0,0 +1,193 @@
+		proxier.ipv4Proxier.OnServiceAdd(service)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnServiceAdd(service)
+	}
+}
+
+// OnServiceUpdate is called whenever modification of an existing
+// service object is observed.
+func (proxier *MetaProxier) OnServiceUpdate(oldService, service *v1.Service) {
+	family := proxier.serviceFamily(service)
+	if family == familyUnknown {
+		family = proxier.serviceFamily(oldService)
+	}
+	switch family {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnServiceUpdate(oldService, service)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnServiceUpdate(oldService, service)
+	}
+}
+
+// OnServiceDelete is called whenever deletion of an existing service
+// object is observed.
+func (proxier *MetaProxier) OnServiceDelete(service *v1.Service) {
+	switch proxier.serviceFamily(service) {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnServiceDelete(service)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnServiceDelete(service)
+	}
+}
+
+// OnServiceSynced is called once all the initial event handlers were
+// called and the state is fully propagated to local cache.
+func (proxier *MetaProxier) OnServiceSynced() {
+	proxier.ipv4Proxier.OnServiceSynced()
+	proxier.ipv6Proxier.OnServiceSynced()
+}
+
+func (proxier *MetaProxier) serviceFamily(service *v1.Service) ipFamily {
+	if service.Spec.ClusterIP == "" {
+		return familyUnknown
+	}
+	if utilnet.IsIPv6String(service.Spec.ClusterIP) {
+		return familyIpv6
+	}
+	return familyIpv4
+}
+
+// OnEndpointsAdd is called whenever creation of new endpoints object
+// is observed.
+func (proxier *MetaProxier) OnEndpointsAdd(endpoints *v1.Endpoints) {
+	switch endpointsFamily(endpoints) {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnEndpointsAdd(endpoints)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnEndpointsAdd(endpoints)
+	}
+}
+
+// OnEndpointsUpdate is called whenever modification of an existing
+// endpoints object is observed.
+func (proxier *MetaProxier) OnEndpointsUpdate(oldEndpoints, endpoints *v1.Endpoints) {
+	family := endpointsFamily(endpoints)
+	if family == familyUnknown {
+		family = endpointsFamily(oldEndpoints)
+	}
+	switch family {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnEndpointsUpdate(oldEndpoints, endpoints)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnEndpointsUpdate(oldEndpoints, endpoints)
+	}
+}
+
+// OnEndpointsDelete is called whenever deletion of an existing
+// endpoints object is observed.
+func (proxier *MetaProxier) OnEndpointsDelete(endpoints *v1.Endpoints) {
+	switch endpointsFamily(endpoints) {
+	case familyIpv4:
+		proxier.ipv4Proxier.OnEndpointsDelete(endpoints)
+	case familyIpv6:
+		proxier.ipv6Proxier.OnEndpointsDelete(endpoints)
+	}
+}
+
+// OnEndpointsSynced is called once all the initial event handlers
+// were called and the state is fully propagated to local cache.
+func (proxier *MetaProxier) OnEndpointsSynced() {
+	proxier.ipv4Proxier.OnEndpointsSynced()
+	proxier.ipv6Proxier.OnEndpointsSynced()
+}
+
+func endpointsFamily(endpoints *v1.Endpoints) ipFamily {
```

Comment:
I don't know that there is such a thing in the long term.  I think it's inevitable that Services become available in both protocols.  It may only be for NodePorts at first, but I predict demand for ClusterIP, too, eventually.

The new EndpointSlice API explicitly allows this (at my insistence, - maybe I am wrong?) and doesn't distinguish v4 from v6 within a slice.  @freehan @robscott

In that case, we'll just want to call both proxiers for all events.  It seems reasonable that the proxiers themselves would filter out wrong-family endpoints anyway.

Maybe we just need to leave that as a comment?

Reply:
Noooo! :anguished:

IMO the decision to have 1service==1family is the most important one in the whole KEP.

It make things so much easier at *every place*, inside and outside k8s, see for instance this [comment in metallb](https://github.com/danderson/metallb/issues/464#issuecomment-523755553).

And I can't see why `NodePort` has to be special in this case. Is it that the node-port will be different for ipv6 and ipv4 since they have different services? If so I can't see the problem. NodePort main usage is for external load-balancers (AWS) and they will most likely setup configs for ipv4 and ipv6 separetely.

Reply:
The POD's are dual-stack and have both addresses, the endpoint addresses for an ipv6 service is taken from the pod's ipv6 addresses and the same for ipv4. Works beautifully IMHO.